Add code for computing n-th roots of unity

Keno opened the pull request:
In many application of finite fields of practical interest, a common
need to compute the an n-th root of unity in this finite field. This
provides (randomized) algorithms for computing either an arbitrary
n-th root or the unique minimum n-th root. Computing the latter is
expensive and of little mathematical significance, but due to the
randomized nature of the algorithm, users may want predictability
in which root gets chosen.

This package seemed like a good place for it.

Keno committed:
Add code for computing n-th roots of unity

In many application of finite fields of practical interest, a common
need to compute the an n-th root of unity in this finite field. This
provides (randomized) algorithms for computing either an arbitrary
n-th root or the unique minimum n-th root. Computing the latter is
expensive and of little mathematical significance, but due to the
randomized nature of the algorithm, users may want predictability
in which root gets chosen.

tkluck commented:
Love it, thank you!

Am I right in thinking that this algorithm would work for any finite field, not just of prime characteristic? If so, maybe we should generalize to `AbstractGaloisField` and not just `PrimeField` (and use `length(F)` instead of `char(F)`).

Keno commented:
It's quite possible that this variant will work, but the literature on primitive root search in non-prime fields is quite extensive and I haven't read through it sufficiently to know that there aren't some corner cases. I also don't have a use case for it that would give me confidence that the code is correct for that case. My preference would be to merge as is and then if and when somebody has a use case for the generalization, we can look into it.

tkluck commented on src/PrimitiveRoots.jl:
```diff
@@ -0,0 +1,63 @@
+"""
+function _rand_root(rng::AbstractRNG, 𝔽::Type{<:PrimeField}, n)
+    # Assumes gcd(char(𝔽) - 1, n) == n
+    r = 𝔽(rand(rng, 1:(char(𝔽) - 1)))
```
There's an implementation for `rand(F)` but for this purpose, it would need two improvements:
 - accept an AbstractRNG
 - have a way to specify nonzero
The former is just good practice and I've wanted the latter for unit testing.

If we find a way to represent this, all of this would work for `AbstractGaloisField`, I think?

Keno replied:
Yes, we could add a `MultiplicativeGroup{𝔽}` wrapper type such that you can have a method for `rand(rng::AbstractRNG, ::Type{MultiplicativeGroup{𝔽}})`.

tkluck commented on src/PrimitiveRoots.jl:
```diff
@@ -0,0 +1,63 @@
+end
+
+"""
+    any_primitive_root(n, 𝔽)
```
typo: `is_primitive_root`

tkluck commented on src/PrimitiveRoots.jl:
```diff
@@ -0,0 +1,63 @@
+
+Obtain some primitive n-th root of unity (not necessarily minimal).
+
+Note that for a primitive n-th root to exist `n` must divide `char(𝔽)`.
```
must divide `char(𝔽) - 1` right?

Keno replied:
yes

tkluck commented on src/PrimitiveRoots.jl:
```diff
@@ -0,0 +1,63 @@
+
+Obtain the (unique) minimal primitive n-th root of unity of the field 𝔽.
+
+Note that for a primitive n-th root to exist `n` must divide `char(𝔽)`.
```
`char(F) - 1` ?

Keno replied:
yes

tkluck commented on src/PrimitiveRoots.jl:
```diff
@@ -0,0 +1,63 @@
+"""
+    minimal_primitive_root(n, 𝔽)
+
+Obtain the (unique) minimal primitive n-th root of unity of the field 𝔽.
```
Maybe it's good to document by what order it's the minimal one.

tkluck commented on src/PrimitiveRoots.jl:
```diff
@@ -0,0 +1,63 @@
+Determine whether `x` is a primitive `n`-th root of unity in 𝔽.
+"""
+is_primitive_root(𝔽::Type{<:PrimeField}, x, n; n_factors = factor(n)) =
+    all(i->!isone(x^(n ÷ i)), keys(n_factors))
```
I trust that you are right, but it's not obvious to me why it's sufficient to only look at `keys(n_factors)` (i.e. the prime divisors) and not all the divisors. Mind adding a comment to explain?

Keno replied:
I'll find a citation.

tkluck reviewed (commented):
Added some more detailed comments.

Also, happy to take your preference and merge this for `PrimeField` first.

tkluck commented:
Oh, there's one more important comment. The thing we are computing here is a "primitive root _of unity_" no? Not a [primitive root](https://en.wikipedia.org/wiki/Primitive_root_modulo_n). These latter ones are actually in use in the [`ZechLog` module](https://github.com/tkluck/GaloisFields.jl/blob/master/src/ZechLog.jl#L86) so it's important to clearly distinguish these. What's a good way to name these functions to make that clear?

Keno commented:
Yes, these's some overlapping confusing terminology. These are `n-th primitive roots of unity in 𝔽p`. A primitive element of 𝔽p (also known as the primitive root of 𝔽p, or the generator of 𝔽p*) is an `p-1-th primitive root of unity in 𝔽p`, so they are related concepts, but the order of the primitive root is implied when only the field is specified. Perhaps the way to resolve this is to just make the `n` argument optional and default to `char(𝔽) - 1`, in which case it matches the primitive element.

Keno commented:
I think I've addressed all the review comments, except the extensions to `AbstractGaloisField` and the `MultiplicativeGroup` abstractions, which are probably better for separate PRs.

tkluck commented:
Thanks @Keno ! I'll make a release soon.
